### PR TITLE
Core implementation of Professional Email upsell content

### PR DESCRIPTION
**Pull request by daledupreez**
#### Changes proposed in this Pull Request

* This PR builds on PR #56671 to flesh out the Professional Email upsell from a content and basic interaction perspective. For now, it includes the following features:
  - Content from i2 design - pcDL0W-I2-p2#comment-963
  - Mobile, tablet, and desktop layouts -- the tablet sizes are a hybrid with the buttons horizontal; please take a look at those sizes
  - Logic to skip adding email
  - Validation of mailbox fields
  - Logic to add the mailbox to the cart _(that doesn't mean it's fully working however -- more below)_
* The following items are _not_ addressed in this PR:
  - We are not wiring in the domain -- we're using a hard-coded `example.com`, which will fail validation when adding the product to the cart
  - We don't have back end support for having some of the mailbox fields be optional - that will be addressed separately

**Feedback requested**
I have not maintained 100% fidelity to the Figma design, with the biggest area of variance being padding and margins, where I've tried to aim for `em`-based values that are pretty close to the designs. I have also incorporated a lot of feedback from @saygunnyc provided via Slack (in this thread: p1633001245007300-slack-C01TJT42PMM) and in this P2 thread: pcDL0W-I2-p2#comment-1012

The other area that needs feedback is the tablet layout, where I've shifted to a columnar layout, but have kept the buttons horizontal. The biggest thing this does is reduce the prominence of the "Why Professional Email?" text -- LMK if you think this text needs to be more prominent or separated.

#### Testing instructions

* As in PR #56671, the first thing we need to do is find an existing receipt ID for one of your sites. Follow these steps to find a site slug and receipt ID:
  - For a site with purchases on it, navigate to Upgrades -> Purchases and then click on the "Billing History" tab
  - For one of the items in the list, click on the "View receipt" link
  - Look at the URL and note the last two `/`-separated pieces -- these are the site slug and receipt ID (in that order)
* Run this branch locally or via the live branch (via the link in [this comment](https://github.com/Automattic/wp-calypso/pull/56690#issuecomment-930988540))
* Manually type the following path into the browser location box: `/checkout/offer-professional-email/:receiptId/:siteSlug` with `:receiptId` and `:siteSlug` replaced by the two values above (which we use in the reverse order for this URL)
* Verify that you see the new content using a domain name of `example.com`
* Try viewing the content at different screen sizes, especially mobile, tablet, and desktop widths - verify that things look OK, are easy to navigate, and match the designs
* Try adding data to the two input fields, including empty/invalid data -- when you enter invalid data, you should see warnings
* Try refreshing the page and immediately clicking the "Add Professional Email" button -- you should see validation errors
* Try entering a valid mailbox name and password, and then click "Add Professional Email"
* Verify that you see an error stating that you don't own the domain -- you can take a look at the network traffic to verify that we're sending valid product data with the `extra`, `quantity`, and `new_quantity` data fields specified

##### Desktop

Screenshot 2021-09-30 at 22 19 46

##### Tablet

Screenshot 2021-09-30 at 22 20 22

##### Mobile

Screenshot 2021-09-30 at 22 20 47
Screenshot 2021-09-30 at 22 20 55

Related to #56671

### CODE REVIEW

File: client/my-sites/checkout/upsell-nudge/professional-email-upsell/index.jsx [hunk 1]
```diff
@@ -1,9 +1,225 @@
...
+	const hasPasswordError =
+		( passwordFieldTouched || showAllErrors ) && null !== mailboxData.password.error;
+
+	const hasBeenValidated = [ mailboxData.mailbox, mailboxData.password ].some(
+		( { error, value } ) => null !== error || '' !== value
+	);
+
+	const optionalMailboxFields = [ 'alternativeEmail', 'name' ];
+
+	const formattedPrice = translate( '{{price/}} /mailbox /month (billed monthly)', {
```
Comment (a8ci18n) [inline client/my-sites/checkout/upsell-nudge/professional-email-upsell/index.jsx]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **17** times:
`translate( '{{price/}} /user /month billed monthly' )` *ES Score: 8*

File: client/my-sites/checkout/upsell-nudge/professional-email-upsell/index.jsx [hunk 2]
```diff
@@ -1,9 +1,225 @@
...
+					</div>
+				</div>
+				<div className="professional-email-upsell__features">
+					<h2>{ translate( 'Why get Professional Email?' ) }</h2>
+					<ul className="professional-email-upsell__feature-list">
+						<ProfessionalEmailFeature>
+							{ translate( "Trusted email address that's truly yours" ) }
+						</ProfessionalEmailFeature>
+						<ProfessionalEmailFeature>
+							{ translate( 'Increase your credibility' ) }
```
Comment (a8ci18n) [inline client/my-sites/checkout/upsell-nudge/professional-email-upsell/index.jsx:206]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **17** times:
`translate( 'Increase your brand credibility' )` *ES Score: 11*

File: client/my-sites/checkout/upsell-nudge/professional-email-upsell/index.jsx [hunk 3]
```diff
@@ -1,9 +1,225 @@
...
+	const optionalMailboxFields = [ 'alternativeEmail', 'name' ];
+
+	const formattedPrice = translate( '{{price/}} /mailbox /month (billed monthly)', {
+		components: {
+			price: <span>{ formatCurrency( productCost ?? 0, currencyCode ) }</span>,
+		},
+		comment: '{{price/}} is the formatted price, e.g. $20',
+	} );
+
+	const isEligibleForFreeTrial = true;
```
Comment (stephanethomas) [inline client/my-sites/checkout/upsell-nudge/professional-email-upsell/index.jsx:74]: How are we going to determine if it is eligible? Do we have to call the backend for this?

File: client/my-sites/checkout/upsell-nudge/professional-email-upsell/index.jsx [hunk 4]
```diff
@@ -1,9 +1,225 @@
...
+		setCartItem( fillInSingleCartItemAttributes( cartItem, productsList ), () =>
+			handleClickAccept( 'accept' )
+		);
+	};
 
-const ProfessionalEmailUpsell = () => {
 	return (
-		<div>
-			<span>Professional Email Upsell placeholder</span>
+		<div className="professional-email-upsell__main">
```
Comment (stephanethomas) [inline client/my-sites/checkout/upsell-nudge/professional-email-upsell/index.jsx]: Do we really need that `professional-email-upsell__main` CSS class? It's not used anywhere.

File: client/my-sites/checkout/upsell-nudge/professional-email-upsell/index.jsx (hunk 3 above)
Comment (daledupreez) [reply client/my-sites/checkout/upsell-nudge/professional-email-upsell/index.jsx:74]: This is part of the work to wire in the correct domain - the data from the `/domains` endpoint includes information about whether a particular domain is eligible for the free trial.

File: client/my-sites/checkout/upsell-nudge/professional-email-upsell/index.jsx (hunk 4 above)
Comment (daledupreez) [reply client/my-sites/checkout/upsell-nudge/professional-email-upsell/index.jsx]: I will remove as part of some other outstanding items that were flagged as feedback on the design post.

**Comment by stephanethomas** [comment]
The code looks solid, and works fine:

PROOF

I'm a bit concerned by the duplication of the form logic as it will be very easy to forgot to update one of those forms whenever we have to fix one of them.

I've noticed that the placeholder doesn't really match the shape of the form, is it something we plan on addressing later?

> €3,75 /mailbox /month (billed monthly) 3 months free

Finally, the `(billed monthly)` mention seems unnecessary to me here.

**Comment by saygunnyc** [comment]
> Finally, the (billed monthly) mention seems unnecessary to me here.

👍 Agree. As we already say /month.

**Comment by daledupreez** [comment]
Thanks for the thorough review, @stephanethomas!

> I've noticed that the placeholder doesn't really match the shape of the form, is it something we plan on addressing later?

This is tricky -- the placeholder comes from the core upsell nudge. I've added the following Asana card to track that: 0-as-1201093507079510/f

**Comment by daledupreez** [comment]
> > Finally, the (billed monthly) mention seems unnecessary to me here.
> 
> 👍 Agree. As we already say /month.

Addressed in my most recent round of commits.

**Comment by saygunnyc** [comment]
One note based on the mobile screenshot shared above.
That would be great to reduce the font size and line height, so "Add Professional Email" fits in a single line and reads better.

Screen Shot 2021-10-01 at 12 05 53

**Comment by stephanethomas** [comment]
> That would be great to reduce the font size and line height, so "Add Professional Email" fits in a single line and reads better.

I would be wary of this kind of refinements as this only works in English, and not in other languages.

**Comment by stephanethomas** [comment]
A heads-up that I've rebased that branch against `trunk`.

**Review by AllTerrainDeveloper — APPROVED** [review]
Looks good to me!

**Evidence of Testing**

Resize for mobile/desktop/tablet
upsell-resize

Validation errors (password is 10 chars length? I've seen it as 12 in othe forms)
image

Error stating that I don't own the domain:
image

**Comment by saygunnyc** [comment]
> I would be wary of this kind of refinements as this only works in English, and not in other languages.

@stephanethomas  Can you further explain this?
I don't understand what line-height has to do with language support.

Currently, the extensive spacing between those 2 lines –"Add Professional" and "Email @example.com" both make the text hard to read and hierarchically confusing as the pricing block now has more proximity to the title than the spacing between the two lines of the header. It gets worse once it's 3 lines. _–They all look separate and not a part of a single header._

Screen Shot 2021-10-05 at 13 13 57

**Comment by stephanethomas** [comment]
Sure thing! I understood from your initial comment that you wanted to reduce the font size so that `Add Professional Email` would fit on one line. I think it's usually better to avoid optimizing the display for English given that there is no guarantee it will work in other languages, e.g. in French this string would be slightly longer (`Ajouter une adresse E-mail Professionnelle`).
You make good points about the height of the line though, and I agree it has nothing to do with the language of the user.

**Comment by stephanethomas** [comment]
I've refined the font size and line height:

screenshot

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/6729332

Thank you @daledupreez for including a screenshot in the description! This is really helpful for our translators.

**Comment by saygunnyc** [comment]
I appreciate it @stephanethomas! 🙌 One last small ask;

Would you mind adding a bit more spacing between the header and pricing for similar reasons? Sharing a quick mockup below.


notes

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/6729332).

**Comment by stephanethomas** [comment]
> Would you mind adding a bit more spacing between the header and pricing for similar reasons? 

This will be addressed in https://github.com/Automattic/wp-calypso/pull/58240 🙂.

**Pull request merged**